cdap-app-fabric/src/test/java/io/cdap/cdap/internal/state/AppStateHandlerTest.java

```diff
@@ -0,0 +1,220 @@
+            new TransactionModules().getInMemoryModules(),
+            new TransactionExecutorModule(),
+            new InMemoryDiscoveryModule(),
+            new MessagingServerRuntimeModule().getInMemoryModules(),
+            new StorageModule(),
+            new AuthorizationTestModule(),
+            new AuthorizationEnforcementModule().getInMemoryModules(),
+            new AuthenticationContextModules().getMasterModule(),
+            new NamespaceAdminTestModule(),
+            new LocalLocationModule(),
+            new PrivateModule() {
+              @Override
+              protected void configure() {
+                bind(MetricsCollectionService.class).to(NoOpMetricsCollectionService.class).in(Scopes.SINGLETON);
+                expose(MetricsCollectionService.class);
+                bind(MetricsSystemClient.class).toInstance(new NoOpMetricsSystemClient());
+                expose(MetricsSystemClient.class);
+              }
+            });
+    appStateStore = new AppStateStore(injector.getInstance(TransactionRunner.class));
+    namespaceAdmin = injector.getInstance(NamespaceAdmin.class);
+    namespaceAdmin.create(new NamespaceMeta.Builder().setName(NAMESPACE_1).build());
+
+    txManager = injector.getInstance(TransactionManager.class);
+    txManager.startAndWait();
+
+    endpoint = "namespaces/" + NAMESPACE_1 + "/apps/" + APP_NAME + "/appids/" + APP_ID + "/states/" + STATE_KEY;
+  }
+
+  @AfterClass
+  public static void teardown() throws Exception {
+    namespaceAdmin.delete(new NamespaceId(NAMESPACE_1));
+
+    if (txManager != null) {
+      txManager.stopAndWait();
+    }
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    // Define all StructuredTable before starting any services that need StructuredTable
+    StoreDefinition.createAllTables(injector.getInstance(StructuredTableAdmin.class));
+
+    NettyHttpService service = new CommonNettyHttpServiceBuilder(CConfiguration.create(), getClass().getSimpleName(),
+            new NoOpMetricsCollectionService())
+            .setHttpHandlers(new AppStateHandler(namespaceAdmin, appStateStore))
+            .build();
+    service.start();
+    config = ClientConfig.builder()
+            .setConnectionConfig(
+                    ConnectionConfig.builder()
+                            .setHostname(service.getBindAddress().getHostName())
+                            .setPort(service.getBindAddress().getPort())
+                            .setSSLEnabled(false)
+                            .build()).build();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+  }
+
+  @Test
+  public void testAppStateSave() throws IOException {
+    // Save state
+    HttpResponse response = executeHttpRequest(HttpMethod.POST, endpoint, STATE_VALUE);
+    Assert.assertEquals(HttpResponseStatus.OK.code(), response.getResponseCode());
+
+    // Cleanup
+    executeHttpRequest(HttpMethod.DELETE, endpoint, null);
```

Comment:
cleanup should be done the tearDown method. Otherwise if something earlier fails, the cleanup won't be performed and other tests could be impacted.